File: platinum/14003_가장_긴_증가하는_부분_수열_5.py
import sys

N = int(sys.stdin.readline())
arr = list(map(int, sys.stdin.readline().split()))
dp = []
LIS = [arr[0]]


def binary_search(target):
    start = 0
    end = len(LIS) - 1

    while start <= end:
        mid = (start + end) // 2

        if LIS[mid] == target:
            return mid
        elif LIS[mid] > target:
            end = mid - 1
        else:
            start = mid + 1
    return start


for i in range(N):
    if LIS[-1] < arr[i]:
        LIS.append(arr[i])
        dp.append((len(LIS) - 1, arr[i]))
    else:
        idx = binary_search(arr[i])
        dp.append((idx, arr[i]))
        LIS[idx] = arr[i]

LIS_LEN = len(LIS)
print(LIS_LEN)

result = [0] * LIS_LEN
idx = LIS_LEN - 1

for i in range(len(dp) - 1, -1, -1):
    if dp[i][0] == idx:
        result[idx] = dp[i][1]
        idx -= 1
print(*result)

File: platinum/14003_가장_긴_증가하는_부분_수열_5.cpp
#include <iostream>
#include <vector>
using namespace std;
 
int N, dp_i = 0;
vector<int> arr, res;
vector<vector<int> > dp;

int mylower_bound(vector<int>& arr, int key){
    int start = 0;
    int end = arr.size() - 1;
    
    int mid;
    
    while(end > start){
        mid = (start + end) / 2;
        if (key > arr[mid]) {
            start = mid + 1;
        }
        else {
            end = mid;
        }
    }
    return end;
}
 
void solve(){
    int cnt = 1;
    res.push_back(arr[0]);
    dp[dp_i][0] = 0;
    dp[dp_i][1] = arr[0];
    dp_i++;

    for (int i = 1; i < N; i++) {
        if(res.back() < arr[i]) {
            res.push_back(arr[i]);
            cnt++;
            
            dp[dp_i][0] = res.size() - 1;
            dp[dp_i][1] = arr[i];
            dp_i++;
        }
        else {
            int idx = mylower_bound(res, arr[i]);
            res[idx] = arr[i];

            dp[dp_i][0] = idx;
            dp[dp_i][1] = arr[i];
            dp_i++;
        }
    }
    
    cout << cnt << endl;

    vector<int> result(cnt);
    int last_idx = cnt - 1;

    for (int i = dp_i - 1; i > -1; i--) {
        if (dp[i][0] == last_idx) {
            
            result[last_idx] = dp[i][1];
            last_idx--;
        }
    }

    for (int i = 0; i < cnt; i++) {
        cout << result[i] << " ";
    }
}
 
int main(){
    cin >> N;
    arr.resize(N + 1);
    dp.resize(N, vector<int>(2, 0));

    for(int i = 0; i < N; i++){
        cin >> arr[i];
    }
    
    solve();
    
    return 0;
}